cookbooks/custom_formatter.rst

```diff
@@ -0,0 +1,465 @@
+Writing a custom Behat formatter
+================================
+
+How to write a custom formatter for Behat?
+
+Introdution
+-----------
+
+Why a custom formatter?
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Behat has three native formatters:
+
+-  **pretty**: the default formatter, which prints every line in green (if a test passes) or red (if it fails),
+-  **progress**: print a "dot" for each test, and a recap of all failing tests at the end,
+-  **junit**: outputs a `junit <https://junit.org/>`__ compatible XML file.
+
+Those are nice, and worked for most of the cases. You can use the "progress" one for the CI, and the "pretty" for development for example.
+
+But you might want to handle differently the output that Behat renders.
+In this cookbook, we will see how to implement a custom formatter for `reviewdog <https://github.com/reviewdog/reviewdog>`__,
+a global review tool that takes input of linters or testers, and that can send "checks" on github, bitbucket or gitlab PR.
+
+Reviewdog can handle `two types of input <https://github.com/reviewdog/reviewdog#input-format>`__:
+
+-  any stdin, coupled with an "errorformat" (a Vim inspired format that can convert text string to machine-readable errors),
+-  a `"Reviewdog Diagnostic Format" <https://github.com/reviewdog/reviewdog/tree/48b25a0aafb8494e751387e16f729faee9522c46/proto/rdf>`__: a JSON with error data that reviewdog can parse.
+
+In my case, I tried parsing Behat's output with errorformat, but I do not know this language, and the multi-line Behat output with "dots" didn't make it easy.
+So I decided to create a custom formatter for Behat.
+
+This way, I will still have Behat's human-readable stdout, and a JSON file written that reviewdog can understand.
+
+Let's dive
+----------
+
+Behat allows us to load "extensions", that can add features to the language. In fact, it is a core functionality to implement PHP functions behind gherkin texts.
+Those extensions are just classes that are loaded by Behat to register configuration and features.
+
+Behat is powered by Symfony: if you know it, you will already know the concepts under the hood, if you don't, that's not a problem and not required to create your extension.
+
+Anatomy of a formatter extension
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A formatter extension requires three things to work:
+
+-  a class that "defines" the extension, to make your extension work with Behat,
+-  a "formatter", that can listen to Behat events, and converts Behat's tests result to anything you want,
+-  an "output printer", that writes the converted data anywhere you want (mainly the stdout, a file or a directory).
+
+Create the extension
+~~~~~~~~~~~~~~~~~~~~
+
+Any Behat extensions must implement ``Behat\Testwork\ServiceContainer\Extension``. Under the hood, it implements Symfony ``CompilerPass``.
```

Comment:
I would remove the mention of CompilerPass here, especially as the extension in this example implements an empty `process` method (which is the only method related to CompilerPass)